packages/sommelier.rb

```diff
@@ -3,112 +3,199 @@
+          : "${SOMMELIER_LOG:=#{CREW_PREFIX}/var/log/sommelier.log}"
+          : "${SOMMELIER_SCALE:=1.0}"
+          : "${XCURSOR_SIZE:=30}"
+          : "${XDG_RUNTIME_DIR:=/var/run/chrome}"
 
           case "$(uname -m)" in
           x86_64|i686)
-            sort_result="$(sort -V <<< "$(uname -r)"$'\\n'"4.16.0" | tail -n1)"
-            if [[ "${sort_result}" == "4.16.0" ]]; then
+            if [[ "$(sort -V <<< "$(uname -r)"$'\\n'"4.16.0" | tail -n1)" == "4.16.0" ]]; then
               MESA_LOADER_DRIVER_OVERRIDE=i965
             fi
           ;;
           esac
 
-          if [ -f "~/.sommelier.env" ]; then
-            source ~/.sommelier.env
+          # locate an unused X display
+          disp=0
+          while [ -S "/tmp/.X11-unix/X${disp}" ]; do
+            ((disp++))
+          done
+          SOMM_X_DISPLAY=":${disp}"
+          # don't export temp vars
+          unset disp
+
+          # locate an unused Wayland display
+          disp='0'
+          # use existing sommelier wayland socket if sommelier already running
+          while [[ -S "${XDG_RUNTIME_DIR}/wayland-${disp}" && "$(< "${XDG_RUNTIME_DIR}/wayland-${disp}.lock")" != 'somm_wl' ]]; do
+            ((disp++))
+          done
+          SOMM_WL_DISPLAY="wayland-${disp}"
+          # don't export temp vars
+          unset disp
+
+          # set X display if not set
+          : "${DISPLAY:=$SOMM_X_DISPLAY}"
+
+          # set WAYLAND_DISPLAY if not set
+          # if WAYLAND_DISPLAY is set to wayland-0 (exo Wayland server), override it
+          if [[ -z "${WAYLAND_DISPLAY}" || ${WAYLAND_DISPLAY} == 'wayland-0' ]]; then
+            WAYLAND_DISPLAY=${SOMM_WL_DISPLAY}
+          fi
+
+          [ -f "~/.sommelier.env" ] && source ~/.sommelier.env
+        EOF
+
+        # sommelierrc
+        # This file via:
+        # crostini: /etc/sommelierrc
+        # https://chromium.googlesource.com/chromiumos/containers/cros-container-guest-tools/+/4329a11adc890173f2385848e7b1bcc89fbd150d/cros-sommelier/sommelierrc
+        @sommelierrc = <<~EOF
+          #!/usr/bin/env bash
+          # Copyright 2019 The Chromium OS Authors. All rights reserved.
+          # Use of this source code is governed by a BSD-style license that can be
+          # found in the LICENSE file.
+          # Startup hook for the sommelier service.  You should not need to modify this
+          # file as customizations may be placed in ~/.sommelierrc instead.
+          #
+          # For documentation, see:
+          # https://chromium.googlesource.com/chromiumos/platform2/+/HEAD/vm_tools/sommelier/
+
+          set -e
+
+          if { command -v xdpyinfo && command -v xrdb; } > /dev/null; then
+            # Set Xft.dpi. Some applications (e.g. Chrome) use this to detect HiDPI.
+            # 1.25x of the DPI reported by xdpyinfo = the DPI reported by xdpyinfo in Crostini
+            DPI="$(( $(xdpyinfo | sed -n -E "/dots per inch/{s|^.* ([0-9]+)x.*$|\\1|g; p}") / 4 * 5 ))"
```

Comment:
This is going to hit a chicken and egg issue.

`sommelierrc` is loaded BEFORE sommelier, and so before a Xserver via Xwayland is running. `xdpyinfo` isn't going to provide info before `xwayland` is running vi `sommelier`, right?

Reply:
Never mind. I was wrong about this.